File: AI-todo/phase2/backend/app/schemas/task.py
"""Task schemas for request/response validation."""

from datetime import date, datetime
from uuid import UUID
from typing import List, Optional

from pydantic import BaseModel, Field

from app.models.task import TaskStatus, PriorityLevel


class TaskCreate(BaseModel):
    """Schema for task creation request."""

    title: str = Field(
        ...,
        min_length=1,
        max_length=200,
        description="Task title (1-200 characters)",
        json_schema_extra={"example": "Buy groceries"},
    )
    description: str | None = Field(
        default=None,
        max_length=1000,
        description="Optional task description",
        json_schema_extra={"example": "Milk, eggs, bread"},
    )
    due_date: date | None = Field(
        default=None,
        description="Optional due date for the task",
        json_schema_extra={"example": "2025-12-31"},
    )
    # Phase V: Advanced features
    priority: PriorityLevel | None = Field(
        default=PriorityLevel.MEDIUM,
        description="Priority level for the task",
    )
    tags: List[str] = Field(
        default_factory=list,
        description="List of tags for the task",
    )
    is_recurring: bool = Field(
        default=False,
        description="Whether this task is part of a recurring pattern",
    )
    recurring_pattern_id: UUID | None = Field(
        default=None,
        description="ID of the recurring pattern this task belongs to",
    )


class TaskUpdate(BaseModel):
    """Schema for task update request (all fields optional)."""

    title: str | None = Field(
        default=None,
        min_length=1,
        max_length=200,
        description="Updated task title",
    )
    description: str | None = Field(
        default=None,
        max_length=1000,
        description="Updated task description",
    )
    status: TaskStatus | None = Field(
        default=None,
        description="Updated task status",
    )
    due_date: date | None = Field(
        default=None,
        description="Updated due date",
    )
    # Phase V: Advanced features
    priority: PriorityLevel | None = Field(
        default=None,
        description="Updated priority level",
    )
    tags: List[str] | None = Field(
        default=None,
        description="Updated list of tags",
    )


class TaskResponse(BaseModel):
    """Schema for task data in responses (includes user_id for backward compatibility)."""

    id: UUID = Field(..., description="Unique task identifier")
    user_id: UUID = Field(..., description="Owner user ID")
    title: str = Field(..., description="Task title")
    description: str | None = Field(..., description="Task description")
    status: TaskStatus = Field(..., description="Task status")
    due_date: date | None = Field(..., description="Due date")
    # Phase V: Advanced features
    priority: PriorityLevel = Field(..., description="Priority level")
    tags: List[str] = Field(..., description="List of tags")
    is_recurring: bool = Field(..., description="Whether this task is part of a recurring pattern")
    recurring_pattern_id: UUID | None = Field(None, description="ID of the recurring pattern this task belongs to")
    is_reminder_sent: bool = Field(..., description="Whether a reminder has been sent")
    parent_task_id: UUID | None = Field(None, description="ID of the parent task (for recurring tasks)")
    next_occurrence_id: UUID | None = Field(None, description="ID of the next occurrence (for recurring tasks)")
    created_at: datetime = Field(..., description="Creation timestamp")
    updated_at: datetime = Field(..., description="Last update timestamp")

    model_config = {"from_attributes": True}


class TaskRead(BaseModel):
    """Schema for task data in responses (excludes user_id for security).

    This schema hides user_id from API responses as the user context
    is already implied by authentication.
    """

    id: UUID = Field(..., description="Unique task identifier")
    title: str = Field(..., description="Task title")
    description: str | None = Field(..., description="Task description")
    status: TaskStatus = Field(..., description="Task status")
    due_date: date | None = Field(None, description="Due date")
    # Phase V: Advanced features
    priority: PriorityLevel = Field(..., description="Priority level")
    tags: List[str] = Field(..., description="List of tags")
    is_recurring: bool = Field(..., description="Whether this task is part of a recurring pattern")
    recurring_pattern_id: UUID | None = Field(None, description="ID of the recurring pattern this task belongs to")
    is_reminder_sent: bool = Field(..., description="Whether a reminder has been sent")
    parent_task_id: UUID | None = Field(None, description="ID of the parent task (for recurring tasks)")
    next_occurrence_id: UUID | None = Field(None, description="ID of the next occurrence (for recurring tasks)")
    created_at: datetime = Field(..., description="Creation timestamp")
    updated_at: datetime = Field(..., description="Last update timestamp")

    model_config = {"from_attributes": True}


class PaginatedTaskResponse(BaseModel):
    """Schema for paginated task list responses."""

    items: list[TaskRead] = Field(..., description="List of tasks")
    total: int = Field(..., description="Total count of matching tasks")
    limit: int = Field(..., description="Page size")
    offset: int = Field(..., description="Current offset")


# Phase V: Advanced features schemas
class SetTaskPriorityRequest(BaseModel):
    """Schema for setting task priority request."""

    priority: PriorityLevel = Field(..., description="New priority level")


class AddTaskTagsRequest(BaseModel):
    """Schema for adding task tags request."""

    tags: List[str] = Field(..., description="List of tags to add")


class RemoveTaskTagsRequest(BaseModel):
    """Schema for removing task tags request."""

    tags: List[str] = Field(..., description="List of tags to remove")


class SearchTasksRequest(BaseModel):
    """Schema for searching tasks request."""

    query: str = Field(..., min_length=1, description="Search query string")
    status: str | None = Field(default=None, description="Filter by status (all, pending, completed)")
    priority: PriorityLevel | None = Field(default=None, description="Filter by priority level")
    tags: List[str] = Field(default_factory=list, description="Filter by tags")
    due_before: date | None = Field(default=None, description="Filter by due date before")
    due_after: date | None = Field(default=None, description="Filter by due date after")
    sort_by: str = Field(default="created_at", description="Field to sort by")
    order: str = Field(default="desc", description="Sort order (asc or desc)")
    page: int = Field(default=1, ge=1, description="Page number")
    per_page: int = Field(default=20, ge=1, le=100, description="Number of items per page")


class CreateRecurringTaskPatternRequest(BaseModel):
    """Schema for creating recurring task pattern request."""

    base_task_title: str = Field(..., min_length=1, max_length=200, description="Base title for recurring tasks")
    base_task_description: str | None = Field(default=None, max_length=1000, description="Base description for recurring tasks")
    pattern_type: str = Field(..., description="Type of recurrence pattern (daily, weekly, monthly, yearly, custom)")
    interval: int = Field(default=1, ge=1, description="Interval between occurrences")
    start_date: date = Field(..., description="Start date for the recurring pattern")
    end_date: date | None = Field(default=None, description="Optional end date for the recurring pattern")
    weekdays: List[int] | None = Field(default=None, description="List of weekdays for weekly patterns (0=Sunday, 6=Saturday)")
    days_of_month: List[int] | None = Field(default=None, description="List of days of month for monthly patterns")

File: AI-todo/phase-5/frontend/src/types/task.ts
/**
 * Task types for the frontend
 * Matches backend API contract from Phase 2 + Phase V advanced features
 */

/**
 * Task status enumeration
 */
export type TaskStatus = 'pending' | 'completed';

/**
 * Priority levels for tasks
 */
export type PriorityLevel = 'low' | 'medium' | 'high';

/**
 * Recurrence pattern types
 */
export type RecurrencePattern = 'daily' | 'weekly' | 'monthly' | 'yearly' | 'custom';

/**
 * Task entity from backend API
 * Note: user_id is NOT included (hidden by backend)
 */
export interface Task {
  id: string;
  title: string;
  description: string | null;
  status: TaskStatus;
  due_date: string | null;  // ISO 8601 date (YYYY-MM-DD)
  created_at: string;   // ISO 8601 datetime
  updated_at: string;   // ISO 8601 datetime
  // Phase V: Advanced features
  priority: PriorityLevel;  // Task priority level
  tags: string[];  // List of tags for the task
  is_recurring: boolean;  // Whether this task is part of a recurring pattern
  recurring_pattern_id: string | null;  // ID of the recurring pattern this task belongs to
  is_reminder_sent: boolean;  // Whether a reminder has been sent
  parent_task_id: string | null;  // ID of the parent task (for recurring tasks)
  next_occurrence_id: string | null;  // ID of the next occurrence (for recurring tasks)
}

/**
 * Create task request payload
 */
export interface TaskCreate {
  title: string;
  description?: string;
  due_date?: string;  // ISO 8601 date (YYYY-MM-DD)
  // Phase V: Advanced features
  priority?: PriorityLevel;  // Task priority level
  tags?: string[];  // List of tags for the task
  is_recurring?: boolean;  // Whether this task is part of a recurring pattern
  recurring_pattern_id?: string;  // ID of the recurring pattern this task belongs to
}

/**
 * Update task request payload (partial)
 */
export interface TaskUpdate {
  title?: string;
  description?: string;
  status?: TaskStatus;
  due_date?: string | null;  // ISO 8601 date (YYYY-MM-DD), null to clear
  // Phase V: Advanced features
  priority?: PriorityLevel;  // Task priority level
  tags?: string[];  // List of tags for the task
}

/**
 * Recurring task pattern entity
 */
export interface RecurringTaskPattern {
  id: string;
  user_id: string;
  base_task_title: string;
  base_task_description: string | null;
  pattern_type: RecurrencePattern;
  interval: number;
  start_date: string;  // ISO 8601 date (YYYY-MM-DD)
  end_date: string | null;  // ISO 8601 date (YYYY-MM-DD)
  weekdays: number[];  // List of weekdays for weekly patterns (0=Sunday, 6=Saturday)
  days_of_month: number[];  // List of days of month for monthly patterns
  created_at: string;   // ISO 8601 datetime
  updated_at: string;   // ISO 8601 datetime
}

/**
 * Create recurring task pattern request payload
 */
export interface RecurringTaskPatternCreate {
  base_task_title: string;
  base_task_description?: string;
  pattern_type: RecurrencePattern;
  interval?: number;
  start_date: string;  // ISO 8601 date (YYYY-MM-DD)
  end_date?: string;  // ISO 8601 date (YYYY-MM-DD)
  weekdays?: number[];  // List of weekdays for weekly patterns (0=Sunday, 6=Saturday)
  days_of_month?: number[];  // List of days of month for monthly patterns
}

/**
 * Set task priority request payload
 */
export interface SetTaskPriorityRequest {
  priority: PriorityLevel;
}

/**
 * Add task tags request payload
 */
export interface AddTaskTagsRequest {
  tags: string[];
}

/**
 * Remove task tags request payload
 */
export interface RemoveTaskTagsRequest {
  tags: string[];
}

/**
 * Search tasks request parameters
 */
export interface TaskSearchParams {
  query: string;
  status?: 'all' | 'pending' | 'completed';
  priority?: PriorityLevel;
  tags?: string[];
  due_before?: string;   // ISO 8601 date
  due_after?: string;    // ISO 8601 date
  sort_by?: 'created_at' | 'title' | 'due_date' | 'priority';
  order?: 'asc' | 'desc';
  page?: number;
  per_page?: number;
}

/**
 * Paginated response wrapper
 */
export interface PaginatedResponse<T> {
  items: T[];
  total: number;
  limit: number;
  offset: number;
}

/**
 * Task list query parameters
 */
export interface TaskQueryParams {
  status?: TaskStatus;
  created_after?: string;   // ISO 8601 date
  created_before?: string;  // ISO 8601 date
  sort?: string;            // Format: "field:direction"
  limit?: number;           // Default: 20, Max: 100
  offset?: number;          // Default: 0
}
